File: backend/src/app/llm/analyze.py
"""Send anonymized log text to a configured LLM provider and return findings.

The request/response shapes differ by provider (OpenAI-style chat completions vs
Anthropic messages); the builders/parsers below are pure and unit-tested, while
``analyze_logs`` wires in the stored key, anonymization and the HTTP call.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any

import httpx
import structlog

from app.llm.anonymize import anonymize
from app.llm.providers import (
    CHAT_ANTHROPIC,
    LLMProvider,
    api_key_setting,
    auth_headers,
    base_url_setting,
    model_setting,
)
from app.settings.store import get_override

log = structlog.get_logger("app.llm")

_TIMEOUT = httpx.Timeout(60.0)
_MAX_INPUT_CHARS = 40_000  # bound tokens/cost; the payload is already capped upstream
# Output budget. On reasoning models (gpt-5.x) this also covers reasoning tokens,
# so too low a value yields an empty answer — keep enough headroom for both.
_MAX_TOKENS = 4000

SYSTEM_PROMPT = (
    "You are a senior network and firewall log analyst for OPNsense/pfSense "
    "systems. Review the log excerpt and report anomalies, errors and "
    "misconfigurations — e.g. ARP flapping or duplicate IPs, interface/driver "
    "errors, failing or restarting services, IPsec tunnel problems, gateway "
    "packet loss, certificate or DNS issues. For each finding give a short "
    "title, a severity (info/warn/critical), the supporting evidence and a "
    "suggested fix. If nothing looks wrong, say so plainly. Be concise and do "
    "not invent issues. Note: IPs/MACs/hostnames may be anonymized."
)


@dataclass(frozen=True)
class AnalyzeResult:
    ok: bool
    provider: str
    model: str
    findings: str
    sent_chars: int
    error: str | None = None


def build_chat_request(
    provider: LLMProvider, key: str, model: str, system: str, user: str
) -> tuple[str, dict[str, str], dict[str, Any]]:
    """URL, headers and JSON body for one chat-completion request. Pure."""
    base = (get_override(base_url_setting(provider.id)) or provider.default_base_url).rstrip("/")
    url = f"{base}{provider.chat_path}"
    headers = auth_headers(provider, key)
    if provider.chat_style == CHAT_ANTHROPIC:
        body: dict[str, Any] = {
            "model": model,
            "max_tokens": _MAX_TOKENS,  # required by Anthropic
            "system": system,
            "messages": [{"role": "user", "content": user}],
        }
    else:
        # OpenAI standardized on max_completion_tokens; newer models (gpt-5.x, o*)
        # reject the legacy max_tokens. OpenRouter accepts it for OpenAI-style too.
        body = {
            "model": model,
            "max_completion_tokens": _MAX_TOKENS,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
        }
    return url, headers, body


def parse_chat_response(chat_style: str, data: dict[str, Any]) -> str:
    """Extract the assistant text from a provider response. Pure."""
    if chat_style == CHAT_ANTHROPIC:
        blocks = data.get("content") or []
        return "".join(b.get("text", "") for b in blocks if b.get("type") == "text").strip()
    choices = data.get("choices") or []
    if not choices:
        return ""
    return (choices[0].get("message") or {}).get("content", "").strip()


async def analyze_logs(provider: LLMProvider, log_text: str) -> AnalyzeResult:
    """Anonymize ``log_text`` and ask ``provider`` to analyze it."""
    key = get_override(api_key_setting(provider.id))
    model = get_override(model_setting(provider.id)) or provider.default_model
    if not key:
        return AnalyzeResult(False, provider.id, model, "", 0, error="No API key configured")
    # anonymize() regex-walks the full log text (up to several hundred KB) —
    # CPU work; a user-triggered analysis must not stall pushes/shells.
    anonymized = (await asyncio.to_thread(anonymize, log_text))[:_MAX_INPUT_CHARS]
    url, headers, body = build_chat_request(provider, key, model, SYSTEM_PROMPT, anonymized)
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(url, headers=headers, json=body)
    except httpx.HTTPError as exc:
        log.warning("llm.analyze_failed", provider=provider.id, error=str(exc))
        return AnalyzeResult(
            False, provider.id, model, "", len(anonymized), error=f"Request failed: {exc}"
        )
    if resp.status_code != 200:
        return AnalyzeResult(
            False,
            provider.id,
            model,
            "",
            len(anonymized),
            error=f"HTTP {resp.status_code}: {resp.text[:300]}",
        )
    findings = parse_chat_response(provider.chat_style, resp.json())
    return AnalyzeResult(True, provider.id, model, findings, len(anonymized))

File: orbit/lib/orbit/llm/analyze.ex
defmodule Orbit.LLM.Analyze do
  @moduledoc """
  AI log analysis — port of llm/providers.py + llm/analyze.py. Invariant 4:
  ONLY anonymized text ever reaches an external LLM (Orbit.LLM.Anonymize,
  which deliberately keeps RFC1918 IPs), capped at 40k chars. Providers
  are a fixed catalog (openai/anthropic/openrouter); key (secret), base
  url and model come from the shared settings registry per provider.

  Test seams: `opts[:req_plug]`, `opts[:settings]`.
  """

  require Logger

  @max_input_chars 40_000
  # Output budget — on reasoning models this also covers reasoning tokens;
  # too low yields an empty answer (analyze.py lesson).
  @max_tokens 4000
  @timeout_ms 60_000

  @system_prompt "You are a senior network and firewall log analyst for OPNsense/pfSense " <>
                   "systems. Review the log excerpt and report anomalies, errors and " <>
                   "misconfigurations — e.g. ARP flapping or duplicate IPs, interface/driver " <>
                   "errors, failing or restarting services, IPsec tunnel problems, gateway " <>
                   "packet loss, certificate or DNS issues. For each finding give a short " <>
                   "title, a severity (info/warn/critical), the supporting evidence and a " <>
                   "suggested fix. If nothing looks wrong, say so plainly. Be concise and do " <>
                   "not invent issues. Note: IPs/MACs/hostnames may be anonymized."

  @providers [
    %{
      id: "openai",
      label: "OpenAI",
      base_url: "https://api.openai.com/v1",
      model: "gpt-5.5",
      auth: :bearer,
      chat_path: "/chat/completions",
      style: :openai
    },
    %{
      id: "anthropic",
      label: "Anthropic",
      base_url: "https://api.anthropic.com",
      model: "claude-opus-4-8",
      auth: :x_api_key,
      chat_path: "/v1/messages",
      style: :anthropic
    },
    %{
      id: "openrouter",
      label: "OpenRouter",
      base_url: "https://openrouter.ai/api/v1",
      model: "openai/gpt-5.5",
      auth: :bearer,
      chat_path: "/chat/completions",
      style: :openai
    }
  ]

  def providers, do: @providers
  def provider(id), do: Enum.find(@providers, &(&1.id == id))

  @doc "Anonymize + cap the text, call the provider, return the findings."
  def analyze_logs(provider_id, log_text, opts \\ []) do
    settings = Keyword.get(opts, :settings, &setting/1)

    with %{} = p <- provider(provider_id) do
      key = to_string(settings.("llm_#{p.id}_api_key") || "")
      model = presence(settings.("llm_#{p.id}_model")) || p.model

      if key == "" do
        {:error, "No API key configured for #{p.label}"}
      else
        anonymized =
          log_text |> Orbit.LLM.Anonymize.anonymize() |> String.slice(0, @max_input_chars)

        base = presence(settings.("llm_#{p.id}_base_url")) || p.base_url
        {url, headers, body} = build_chat_request(p, base, key, model, anonymized)
        post(p, url, headers, body, model, opts)
      end
    else
      nil -> {:error, "unknown provider"}
    end
  end

  @doc "URL, headers and JSON body for one chat request. Pure."
  def build_chat_request(p, base, key, model, user_text) do
    url = String.trim_trailing(base, "/") <> p.chat_path

    headers =
      case p.auth do
        :bearer -> [{"authorization", "Bearer #{key}"}]
        :x_api_key -> [{"x-api-key", key}, {"anthropic-version", "2023-06-01"}]
      end

    body =
      case p.style do
        :anthropic ->
          %{
            model: model,
            max_tokens: @max_tokens,
            system: @system_prompt,
            messages: [%{role: "user", content: user_text}]
          }

        :openai ->
          # max_completion_tokens: newer OpenAI models reject legacy max_tokens.
          %{
            model: model,
            max_completion_tokens: @max_tokens,
            messages: [
              %{role: "system", content: @system_prompt},
              %{role: "user", content: user_text}
            ]
          }
      end

    {url, headers, body}
  end

  @doc "Extract the assistant text from a provider response. Pure."
  def parse_chat_response(:anthropic, data) do
    (data["content"] || [])
    |> Enum.filter(&(&1["type"] == "text"))
    |> Enum.map_join("", & &1["text"])
    |> String.trim()
  end

  def parse_chat_response(:openai, data) do
    case data["choices"] do
      [first | _] -> String.trim(get_in(first, ["message", "content"]) || "")
      _ -> ""
    end
  end

  defp post(p, url, headers, body, model, opts) do
    base_opts = [
      url: url,
      headers: headers,
      json: body,
      receive_timeout: @timeout_ms,
      retry: false
    ]

    req_opts =
      case Keyword.get(opts, :req_plug, Application.get_env(:orbit, :llm_req_plug)) do
        nil -> base_opts
        plug -> Keyword.put(base_opts, :plug, plug)
      end

    case Req.post(req_opts) do
      {:ok, %{status: 200, body: data}} when is_map(data) ->
        {:ok, %{provider: p.id, model: model, findings: parse_chat_response(p.style, data)}}

      {:ok, %{status: status, body: body}} ->
        {:error, "HTTP #{status}: #{String.slice(inspect(body), 0, 300)}"}

      {:error, error} ->
        Logger.warning("llm.analyze_failed provider=#{p.id} error=#{Exception.message(error)}")
        {:error, "Request failed: #{Exception.message(error)}"}
    end
  end

  defp setting(key) do
    Orbit.Settings.effective(key)
  rescue
    _ -> ""
  end

  defp presence(v) do
    case String.trim(to_string(v || "")) do
      "" -> nil
      s -> s
    end
  end
end
